### DISCUSSION TITLE
Custom layer with renderWorldCopies

### DISCUSSION THREAD

**Discussion by tangzijun** [maplibre/maplibre-gl-js#4614] [Q&A]
**Question**
How to implement renderWorldCopies in a custom layer? This is not implemented in the official examples. Does anyone have any ideas for implementation?

**Links to related documentation**
https://maplibre.org/maplibre-gl-js/docs/examples/custom-style-layer/

**Comment by HarelM** [comment]
Can you elaborate what problem you are trying to solve? Or what limitation did you find? 
Unfortunately, I couldn't understand from the above description...
Also, try one of the AI tools, maybe they can help, I use them a lot 😃

**Comment by tangzijun** [comment]
In this official example, custom layers are not rendered in multiple world copies.

https://maplibre.org/maplibre-gl-js/docs/examples/custom-style-layer/


![image](https://github.com/user-attachments/assets/04a3a1b0-a522-4219-9722-489ac7fa295a)

![image](https://github.com/user-attachments/assets/069ca0ed-8d4c-4a6e-b805-7cf418d1dacc)


> Can you elaborate what problem you are trying to solve? Or what limitation did you find? Unfortunately, I couldn't understand from the above description... Also, try one of the AI tools, maybe they can help, I use them a lot 😃

**Comment by tangzijun** [comment]
The layers that come with the framework, such as the image layer, will render multiple copies in the world copy. How can I achieve this in a custom layer?

**Comment by HarelM** [comment]
In order to solve this I added 2 more buffers with coordinates that are unwrapped (+360, -360 longitude) and drew them as well.
I'm not sure this is the optimal solution, but custom layer is basically how you decide to draw on the canvas given the relevant input.

**Comment by tangzijun** [comment]
I think I solved this problem，There are 3 steps to solve this problem

1. Get the offset of the world copy
```
private getWorldOffsets() {
  const bounds = this.map.getBounds();
  const west = bounds.getWest();
  const east = bounds.getEast();

  const x1 = 0 - Math.ceil((Math.abs(west) - 180) / 360);
  const x2 = Math.ceil((Math.abs(east) - 180) / 360);

  const worldOffsets = [];
  for (let i = x1; i <= x2; i++) {
    worldOffsets.push(i);
  }
  return worldOffsets;
}
```
This will get an array worldOffsets: […,-1,0,1…]


2. Cycle Rendering
```
worldOffsets.forEach(offset => {
     //Start rendering and add the offset to the shader variable: u_worldXOffset
  });
```


3. Implementing the offset in the shader
```
#version 300 es

in vec4 position;
in vec2 texcoord;
uniform mat4 u_matrix;
uniform float u_worldXOffset;

void main() {
    gl_Position = u_matrix * (position + vec4(u_worldXOffset , 0.0 , 0.0 , 0.0));
…
}
```

**Comment by carderne** [comment]
There was an issue at the Mapbox repo about this here:
https://github.com/mapbox/mapbox-gl-js/issues/9759

Possibly helpful.

**Comment by su-noh** [comment]
The cause of this problem is due to rendering the map in mercator method. Since mapbox and maplibre are tile-based maps, the above problem does not occur by rendering all the tiles currently visible, but as far as I know, the tile information is not provided in the custom layer. 

This issue shares a similar context to what I asked above in this discussion
[https://github.com/maplibre/maplibre-gl-js/discussions/5293](https://github.com/maplibre/maplibre-gl-js/discussions/5293)
